### PR TITLE
Add Spanish translations to Financial Well-Being Scale Django app

### PR DESCRIPTION
**Pull request by marteki**
This adds an iteration on the Django app version of the Financial Well-Being Scale (https://www.consumerfinance.gov/consumer-tools/financial-well-being/) to include its Spanish version.

## Additions

- A `TranslatedTemplateView` view to direct users to the proper URL for their chosen language 
- Links for users to switch between English and Spanish on most pages of the app
- Spanish translations for the Financial Well-Being Scale added to `django.po`

## Testing

1. Load this branch on your machine, and visit http://localhost:8000/consumer-tools/financial-well-being/
2. Fill out the entire questionnaire, checking that the JavaScript and data flow for the default English are working as expected. Check the `/financial-well-being/about/` and `/financial-well-being/error/` slugs as well.
3. Return to the start of the questionnaire, and click the link under the hero to view the page in Spanish: "Ver página en español". Fill out the entire questionnaire in Spanish, checking that the JavaScript and data flow for Spanish are working as expected. Check the `/financial-well-being/about/es/` and `/financial-well-being/error/es/` slugs as well.
4. Return to the start of the questionnaire, and click the link under the hero to view the page in English again: "View page in English".
5. Manually add `/es/` to the end of the default start page for the tool. Check that the page still switches to the Spanish translation.

### Start page

fwb-home-english

fwb-home-spanish

### Results page

fwb-results-english

fwb-results-spanish

### About page

fwb-about-english

fwb-about-spanish

### Error page

fwb-error-english

fwb-error-spanish

## Notes

- There isn't a link to switch between English and Spanish on the `/results/` page, as the user's data doesn't flow through a change in the views. You can see what would happen by manually adding `/es/` to the end of the URL after filling out the questionnaire.

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [x] Project documentation has been updated
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Browsers

- [x] Chrome on desktop
- [x] Safari on macOS

### Other

- [x] Flexible from small to large screens
- [x] No linting errors or warnings
- [x] JavaScript tests are passing

### CODE REVIEW

File: cfgov/wellbeing/urls.py [hunk 1]
```diff
@@ -10,17 +11,60 @@
...
+        ),
+        name='fwb_about_es'
+    ),
+    url(
+        r'^error/es/$',
+        TranslatedTemplateView.as_view(
+            template_name='wellbeing/error.html',
+            language='es'
+        ),
+        name='fwb_error_es'
```
Comment (willbarton) [inline cfgov/wellbeing/urls.py:68]: I think the `re_path` removal above was a casualty of merging. We should replace the urls below with re_path, since that's what we standardized on going forward for Django 2.

```suggestion
    re_path(
        r'^$',
        TranslatedTemplateView.as_view(
            template_name='wellbeing/home.html'
        ),
        name='fwb_home_en'
    ),
    re_path(
        r'^results/$',
        ResultsView.as_view(),
        name='fwb_results_en'
    ),
    re_path(
        r'^about/$',
        TranslatedTemplateView.as_view(
            template_name='wellbeing/about.html'
        ),
        name='fwb_about_en'
    ),
    re_path(
        r'^error/$',
        TranslatedTemplateView.as_view(
            template_name='wellbeing/error.html'
        ),
        name='fwb_error_en'
    ),

    re_path(
        r'^es/$',
        TranslatedTemplateView.as_view(
            template_name='wellbeing/home.html',
            language='es'
        ),
        name='fwb_home_es'
    ),
    re_path(
        r'^results/es/$',
        ResultsView.as_view(language='es'),
        name='fwb_results_es'
    ),
    re_path(
        r'^about/es/$',
        TranslatedTemplateView.as_view(
            template_name='wellbeing/about.html',
            language='es'
        ),
        name='fwb_about_es'
    ),
    re_path(
        r'^error/es/$',
        TranslatedTemplateView.as_view(
            template_name='wellbeing/error.html',
            language='es'
        ),
        name='fwb_error_es'
```

File: cfgov/wellbeing/urls.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
-from django.views.generic import TemplateView
+from django.conf.urls import url
 
+from core.views import TranslatedTemplateView
 from wellbeing.views import ResultsView
```
Comment (willbarton) [inline cfgov/wellbeing/urls.py:4]: We'll need to import `re_path` in Django 2 or import `url` as `re_path` in Django 1.11.

```suggestion

from core.views import TranslatedTemplateView
from wellbeing.views import ResultsView


try:
    from django.urls import re_path
except ImportError:
    from django.conf.urls import url as re_path
```

File: cfgov/wellbeing/jinja2/wellbeing/results.html [hunk 3]
```diff
@@ -409,35 +404,45 @@ <h3>Take control of day-to-day money management</h3>
...
             <div class="content-l_col content-l_col-1-2">
-                <h3>Get on track for your financial future</h3>
+                <h3>{{ _('Get on track for your financial future') }}</h3>
+                {% if (current_language == 'en') %}
                 <ul>
                     <li>
                         <b>Look toward home ownership.</b>
-                        Consult our
-                        <a href="/owning-a-home/">Buying a House</a>
+			                  Consult our <a href="/owning-a-home/">Buying a House</a>
```
Comment (Scotchester) [inline cfgov/wellbeing/jinja2/wellbeing/results.html:434]:
```suggestion
                        Consult our <a href="/owning-a-home/">Buying a House</a>
```

File: cfgov/wellbeing/jinja2/wellbeing/results.html [hunk 4]
```diff
@@ -409,35 +404,45 @@ <h3>Take control of day-to-day money management</h3>
...
-                           href="https://files.consumerfinance.gov/f/documents/201705_cfpb_My-New-Money-Goal.pdf">
-                            <span class="a-link_text">worksheet</span>
-                            {{ svg_icon('download') }}
-                        </a>
+                        <a class="a-link a-link__icon"
+                        href="https://files.consumerfinance.gov/f/documents/201705_cfpb_My-New-Money-Goal.pdf">
+                        <span class="a-link_text">
+                          worksheet
+                        </span>
+                            {{ svg_icon('download') }}</a>
```
Comment (Scotchester) [inline cfgov/wellbeing/jinja2/wellbeing/results.html:445]:
```suggestion
                        <a class="a-link
                                  a-link__icon"
                           href="https://files.consumerfinance.gov/f/documents/201705_cfpb_My-New-Money-Goal.pdf">
                            <span class="a-link_text">worksheet</span>
                            {{ svg_icon('download') }}
                        </a>
```

File: cfgov/wellbeing/jinja2/wellbeing/results.html [hunk 5]
```diff
@@ -461,89 +468,130 @@ <h3>Get on track for your financial future</h3>
                         and their effect on your cost of borrowing.
                     </li>
                 </ul>
+                {% endif %}
+                {% if (current_language == 'es') %}
```
Comment (Scotchester) [inline cfgov/wellbeing/jinja2/wellbeing/results.html:472]:
```suggestion
                {% else %}
```

File: cfgov/wellbeing/jinja2/wellbeing/results.html [hunk 6]
```diff
@@ -461,89 +468,130 @@ <h3>Get on track for your financial future</h3>
...
+                <ul>
+                    <li>
+                        <b>¿Está considerando comprar una casa?</b>
+			                  Tenemos <a href="/es/prepararse-para-comprar-casa/">
+                          recursos para ayudarle</a>.
```
Comment (Scotchester) [inline cfgov/wellbeing/jinja2/wellbeing/results.html:477]:
```suggestion
                        Tenemos <a href="/es/prepararse-para-comprar-casa/">recursos para ayudarle</a>.
```

File: cfgov/wellbeing/jinja2/wellbeing/results.html [hunk 7]
```diff
@@ -461,89 +468,130 @@ <h3>Get on track for your financial future</h3>
...
+                    </li>
+                    <li>
+                        <b>Desarrolle un plan de acción para alcanzar sus metas.</b>
+                        Use esta
+                        <a href="https://pueblo.gpo.gov/CFPBLibs/CFPBLibsPubs.php?PubID=13757">
+                          hoja de cálculo
+                        </a>
+                        , o herramienta para hacer un presupuesto.
```
Comment (Scotchester) [inline cfgov/wellbeing/jinja2/wellbeing/results.html:485]:
```suggestion
                        <a href="https://pueblo.gpo.gov/CFPBLibs/CFPBLibsPubs.php?PubID=13757">hoja de cálculo</a>,
                        o herramienta para hacer un presupuesto.
```

File: cfgov/wellbeing/jinja2/wellbeing/results.html [hunk 8]
```diff
@@ -461,89 +468,130 @@ <h3>Get on track for your financial future</h3>
...
-                See
+                <b>A nonprofit credit counselor</b> can help you set up a
+                reasonable repayment plan for credit cards and other consumer
+                debt. See
                 <a href="/ask-cfpb/how-do-i-choose-which-credit-counselor-is-right-for-me-en-1355/">
-                    tips for choosing a credit counselor</a>.
+                  tips for choosing a credit counselor</a>.
+            </li>
+            {% endif %}
+            {% if (current_language == 'es') %}
```
Comment (Scotchester) [inline cfgov/wellbeing/jinja2/wellbeing/results.html:523]:
```suggestion
            {% else %}
```

File: cfgov/wellbeing/jinja2/wellbeing/results.html [hunk 9]
```diff
@@ -461,89 +468,130 @@ <h3>Get on track for your financial future</h3>
...
+                <a href="/find-a-housing-counselor/">{{ _('Find one in your area.') }}</a>
             </li>
             <li>
-                <b>A financial coach</b>
-                can assess your situation and work with you over time to resolve issues and get ahead.
+                <b>{{ _('A financial coach') }}</b>
+                {{ _('can assess your situation and work with you over time to resolve issues and get ahead.') }}
                 <a href="/about-us/blog/the-launch-of-the-cfpb-financial-coaching-initiative/">
-                    Find out about the CFPB coaching initiative.
+                {{ _('Find out about the CFPB coaching initiative.') }}
```
Comment (Scotchester) [inline cfgov/wellbeing/jinja2/wellbeing/results.html:542]:
```suggestion
                    {{ _('Find out about the CFPB coaching initiative.') }}
```

File: cfgov/wellbeing/urls.py [hunk 10]
```diff
@@ -1,8 +1,14 @@
-from django.views.generic import TemplateView
 
+from core.views import TranslatedTemplateView
```
Comment (Scotchester) [inline cfgov/wellbeing/urls.py:2]:
```suggestion
from core.views import TranslatedTemplateView
```

File: cfgov/wellbeing/urls.py [hunk 11]
```diff
@@ -1,8 +1,14 @@
...
 from wellbeing.views import ResultsView
 
 
+try:
+    from django.urls import re_path
+except ImportError:
+    from django.conf.urls import url as re_path
+
+
```
Comment (Scotchester) [inline cfgov/wellbeing/urls.py:11]: This already got added in a separate Django 2 support PR.

```suggestion
```

File: cfgov/wellbeing/views.py [hunk 12]
```diff
@@ -1,19 +1,20 @@
...
-class ResultsView(TemplateView):
+class ResultsView(TranslatedTemplateView):
     template_name = 'wellbeing/results.html'
 
     # for simplicity, answers below are reordered in order of worst to best
     QUESTIONS = [
         {
             # Question 1
-            'question': 'I could handle a major unexpected expense',
+            'question': _('I could handle a major unexpected expense'),
```
Comment (Scotchester) [inline cfgov/wellbeing/views.py:17]: Do the rest of the questions (and the answers) in this data structure need to be translated?

Does this one even need to be? Looks like the values in this structure get translated when being output on the results template: https://github.com/cfpb/cfgov-refresh/pull/5668/files#diff-13377cbaabf59c01cfdebe7d4f3e2a1dR355-R360

File: cfgov/wellbeing/jinja2/wellbeing/results.html [hunk 13]
```diff
@@ -230,36 +230,32 @@
...
             'warning',
             true,
-            'We never collect or store answers to the financial well-being questionnaire.',
-            'If you would like to see your score and answers on the page below,
-            please start with the <a href="../">questionnaire</a> and enter your responses.' | safe
+            _('We never collect or store answers to the financial well-being questionnaire.'),
+	          _('If you would like to see your score and answers on the page below, please start with the <a href="../">questionnaire</a> and enter your responses.') | safe
         ) }}
     </div>
 {% endif %}
```
Comment (Scotchester) [inline cfgov/wellbeing/jinja2/wellbeing/results.html:248]: This notification that appears if you manage to go to the results page without having submitted the form is not being translated to Spanish.

File: cfgov/wellbeing/jinja2/wellbeing/results.html [hunk 14]
```diff
@@ -461,89 +468,130 @@ <h3>Get on track for your financial future</h3>
...
             true,
-            'We never collect or store answers to the financial well-being questionnaire.',
-            'If you would like to see your score and answers on the page below,
-            please start with the <a href="../">questionnaire</a> and enter your responses.'
+            _('We never collect or store answers to the financial well-being questionnaire.'),
+            _('If you would like to see your score and answers on the page below, please start with the <a href="../">questionnaire</a> and enter your responses.')
         ) }}
-    </div>
+        </div>
     {% endif %}
```
Comment (Scotchester) [inline cfgov/wellbeing/jinja2/wellbeing/results.html:566]: This one is not being translated, either.

File: cfgov/wellbeing/jinja2/wellbeing/results.html [hunk 15]
```diff
@@ -582,21 +630,22 @@ <h4>Compare by</h4>
 
             <figcaption class="comparison-chart_source">
                 <dl>
-                    <dt>Source:</dt>
+                    <dt>{{ _('Source:') }}</dt>
                     <dd>
-                        <a href="/data-research/research-reports/financial-well-being-america/">Financial Well-Being in America report</a> |
-                        <a href="https://files.consumerfinance.gov/f/documents/cfpb_well-being_score-comparison-data.csv">Download CSV file</a>
+                        <a href="/data-research/research-reports/financial-well-being-america/">{{ _('Financial Well-Being in America report') }}</a> |
```
Comment (Scotchester) [inline cfgov/wellbeing/jinja2/wellbeing/results.html:635]: This link text is not being translated.

File: cfgov/wellbeing/views.py (hunk 12 above)
Comment (marteki) [reply cfgov/wellbeing/views.py:17]: @willbarton and I tried to apply translations in the data structure, but in his words:
"at the point where QUESTIONS is defined, english (the default language) is activated"
"the url-dependent language is only activated after get_context_data is called, but becuase those questions are defined on the view, they are evaluated when the server starts up"
So we needed to find a different way of exposing the translations

They were added in `django.po` with the following comment: "# Manually added translations from within wellbeing/jinja2/wellbeing/views.py"

File: cfgov/wellbeing/jinja2/wellbeing/results.html (hunk 13 above)
Comment (marteki) [reply cfgov/wellbeing/jinja2/wellbeing/results.html:248]: Just saw this yesterday, didn't get a chance to fix it yet. If it's not being picked up by `makemessages`, I'll probably take the same approach I used (with Will's blessing) for the text contained inside the views.

File: cfgov/wellbeing/jinja2/wellbeing/results.html (hunk 15 above)
Comment (marteki) [reply cfgov/wellbeing/jinja2/wellbeing/results.html:635]: Oddly enough, that is the desired translation for the stakeholders.

Screen Shot 2020-04-22 at 9 12 44 AM

File: cfgov/wellbeing/urls.py (hunk 11 above)
Comment (marteki) [reply cfgov/wellbeing/urls.py:11]: @willbarton asked me to re-add it in a separate commit suggestion above, because he thought it was removed by mistake in one of the rebases? I just changed it yesterday.

**Comment by marteki** [comment]
@higs4281 

> On the Spanish pages, some elements of the header and footer are not picking up sitewide translations available to Spanish Wagtail pages that use the CFGOVPage base, such as Blogs and Ask CFPB pages. Those pages check the Wagtail `page` object to determine language, and so pick up a Spanish CFPB logo and header menu and similar menu choices in the footer.

Ah, cool to learn about this ability! I defer to @willbarton as the one who has primarily been working on the Django side of this with me to determine complexity/LOE, but if it's not super simple I will wait to change that for later (so the stakeholders can get the version they okayed live on the site by Friday as expected).

**Comment by marteki** [comment]
@Scotchester I've been getting the exact same sort of Chrome rendering errors **starting after our consult about the expandable issues**. The only way I can fix it on my localhost is by **rerunning `frontend.sh`**, if that provides any insight on how to fix it properly and permanently.

**Comment by Scotchester** [comment]
> @Scotchester I've been getting the exact same sort of Chrome rendering errors **starting after our consult about the expandable issues**. The only way I can fix it on my localhost is by **rerunning `frontend.sh`**, if that provides any insight on how to fix it properly and permanently.

I got a Chrome update this morning that seems to have fixed it.

File: cfgov/wellbeing/jinja2/wellbeing/results.html (hunk 15 above)
Comment (Scotchester) [reply cfgov/wellbeing/jinja2/wellbeing/results.html:634]: Hmm… I guess maybe because the report isn't translated? I might remove the translation wrapper, then, so future devs don't expect it to be translated when they see the code, and then it's not translated on-page.

File: cfgov/wellbeing/views.py (hunk 12 above)
Comment (Scotchester) [reply cfgov/wellbeing/views.py:17]: I don't think they're needed, because they get translated when actually accessed for output in the template.

File: cfgov/wellbeing/urls.py (hunk 11 above)
Comment (Scotchester) [reply cfgov/wellbeing/urls.py:10]: Yeah, I think he was partially mistaken in that suggestion he made. The `from django.conf.urls import url` on L1 needed to be removed, but the block he suggested adding was already there further down.

Regardless of what happened, it definitely doesn't need to be in there twice now.

Comment (willbarton) [reply cfgov/wellbeing/urls.py:10]: Yeah, I think I missed that it was already in that file (that wasn't obvious in the diff). Sorry about that!

File: cfgov/wellbeing/views.py (hunk 12 above)
Comment (willbarton) [reply cfgov/wellbeing/views.py:17]: @Scotchester they are wrapped in `_()` in both places for complicated reasons.

By adding `_()` around them in the view code, they get pulled into the `messages.po` file by Django's `makemessages`. We could add it by hand there, but this way Django's `makemessages` will give us line numbers, and update them when `makemessages` is run.

*BUT* at the time when the view is evaluated, the language that's "active" is English, so they are always English. Spanish is only activated when the context is fetched for the page.

By wrapping the values in the template with `_()` we ensure that the string gets translated to Spanish after Spanish is activated.

TLDR: The `_()` right here does nothing to translate the string. It's only here to make sure the string gets auto-populated in `messages.po`.

Comment (Scotchester) [reply cfgov/wellbeing/views.py:17]: @willbarton Two questions then:

1. Reiterating the first question of this thread: "Do the rest of the questions (and the answers) in this data structure need to be translated?"
2. If so, why am I seeing them translated, anyway? I suspect because all of these strings all already existing in `messages.po` from being `_()` wrapped when they appear in the `home.html` template?

**Comment by willbarton** [comment]
@higs4281 @marteki Good discovery of what this wasn't working!

> ```jinja2
> {% if current_language %}
> {% set language = current_language %}
> {% else %}
> {% set language = page.language if page and page.language else 'en' %}
> {% endif %}
> ```

I don't love the conditional wrapped around a ternary. If this is all it takes, we could include it here. I'm happy to push a change to make it happen, @marteki. I might try to find a different way to handle these conditionals though.

File: cfgov/wellbeing/views.py (hunk 12 above)
Comment (willbarton) [reply cfgov/wellbeing/views.py:17]: Oh!

I wonder — @marteki and I went back and forth on this a couple times. @marteki did we have all of them wrapped at one time? Or did you add them by hand?

`makemessages` doesn't *evaluate* the view and template, so `_(my_variable)` won't get added to `messages.po` by it.

Comment (willbarton) [reply cfgov/wellbeing/views.py:17]: I do think we should wrap all of these strings so that the location they occur is recorded, but that's for human/developer/designer-friendliness, not for technical reasons.

Comment (marteki) [reply cfgov/wellbeing/views.py:17]: @willbarton - I wrapped them all originally in the view. They weren't showing up in `messages.po` at all. [The Django docs say this](https://docs.djangoproject.com/en/3.0/topics/i18n/translation/):

>(The caveat with using variables or computed values, as in the previous two examples, is that Django’s translation-string-detecting utility, django-admin makemessages, won’t be able to find these strings. More on makemessages later.)

We talked as a result. I added them by hand to `messages.po`, manually entering them with line numbers. At one point, I removed all the wraps from the view, but maybe some crept their way back in.

Comment (willbarton) [reply cfgov/wellbeing/views.py:17]: Hmm, okay. I guess let's remove the last couple from the view?

**Comment by willbarton** [comment]
@higs4281 @marteki I've pushed a commit to include the `current_language` in the base template. I replaced the ternary altogether with an `if`/`elif`/`else`.

**Comment by marteki** [comment]
Updates made for the attention of @Scotchester and @willbarton. Two commits added to:

- Translate the notification if the results page is reached without filling out the questionnaire
- Remove the translation tags from around a report title that will not be translated from English
- Clean up the single unused translation tag accidentally left in the view

File: cfgov/wellbeing/jinja2/wellbeing/results.html (hunk 13 above)
Comment (marteki) [reply cfgov/wellbeing/jinja2/wellbeing/results.html]: Addressed in newest commits.

File: cfgov/wellbeing/views.py (hunk 12 above)
Comment (marteki) [reply cfgov/wellbeing/views.py]: Addressed in newest commits.

File: cfgov/wellbeing/jinja2/wellbeing/results.html (hunk 15 above)
Comment (marteki) [reply cfgov/wellbeing/jinja2/wellbeing/results.html]: Addressed in newest commits.

File: cfgov/wellbeing/jinja2/wellbeing/results.html (hunk 14 above)
Comment (marteki) [reply cfgov/wellbeing/jinja2/wellbeing/results.html:571]: Addressed in newest commits.

Comment (Scotchester) [reply cfgov/wellbeing/jinja2/wellbeing/results.html:571]: This one is still not being translated. I think you're going to have to do the same thing you did with the `current_language` conditional in the first of this kind of notification at the top of this page.

**Comment by willbarton** [comment]
@marteki the view changes look good to me. Do you want me to push a linting fix (the `_` import in the `views.py` can be removed)?

File: cfgov/wellbeing/jinja2/wellbeing/results.html (hunk 14 above)
Comment (marteki) [reply cfgov/wellbeing/jinja2/wellbeing/results.html]: Thanks for the sharp eyes! Fixed.
